Indexing/InductionFormulaIndex.hpp

```diff
@@ -0,0 +1,87 @@
+/*
+ * This file is part of the source code of the software program
+ * Vampire. It is protected by applicable
+ * copyright laws.
+ *
+ * This source code is distributed under the licence found here
+ * https://vprover.github.io/license.html
+ * and in the source directory
+ */
+/**
+ * @file InductionFormulaIndex.hpp
+ * Defines class InductionFormulaIndex.
+ */
+
+
+#ifndef __InductionFormulaIndex__
+#define __InductionFormulaIndex__
+
+#include "Forwards.hpp"
+
+#include "Lib/DHMap.hpp"
+#include "Lib/Hash.hpp"
+#include "Lib/Stack.hpp"
+
+#include "Kernel/Clause.hpp"
+#include "Kernel/Substitution.hpp"
+
+namespace Inferences {
+  struct InductionContext;
+}
+
+namespace Indexing {
+
+using namespace Lib;
+using namespace Kernel;
+
+struct SecondaryStackHash {
+  static unsigned hash(const Stack<LiteralStack>& s) {
+    unsigned res = FNV32_OFFSET_BASIS;
+    typename Stack<LiteralStack>::ConstIterator it(s);
+    while (it.hasNext()) {
+      res += it.next().length();
+    }
+    return res;
+  }
+};
+
+class InductionFormulaIndex
+{
+public:
+  /** Stores clausified induction formulas,
+   * each associated with a substitution that
+   * needs to be applied on conclusion literals
+   * to get the complementary literals stored
+   * in a matching InductionContext.
+   */
+  struct Entry {
+    void add(ClauseStack&& cls, Substitution&& subst) {
+      if (cls.isEmpty()) {
+        return;
+      }
+      // Each clause is used immediately upon creation but
+      // their memory will be reused by Vampire if their
+      // store is NONE and all their descendants are deleted
+      // due to simplifications, so we change them to active.
+      for (const auto& cl : cls) {
```

Comment:
My experience so far is that setting the store to active is the least harmful way of saying that I want to retain this clause in the long run, otherwise leaving it to none can result in its memory being reused. Maybe introducing a new store would be better?

Reply:
Guys, look at:

```
bool Clause::shouldBeDestroyed()
{
  return (_store == NONE) && _refCnt == 0 &&
    !isFromPreprocessing();
}
```

while _store has meaning for the saturation algorithm, reference counting is exactly for the purpose if keeping things alive, if they are still needed. 

Couldn't you simply replace your ClauseStack in `_st` by `RCClauseStack`? Or just do the increase/decrease yourself, where appropriate. Ideally, using some from of [RAII](https://en.cppreference.com/w/cpp/language/raii) - @MichaelRawson will know how! :)

Reply:
Yeah, that's a good idea now that I know this! Then I would rather increase the refcount here instead of setting the store or using `RCClauseStack` because otherwise either I lose some good features of `Stack` (like the subscript operator) or conversion is needed. And we never remove clauses from this index, so no need for decrease of refcount. What do you think?

Reply:
Yes, sure, just increase the refcount on insertion!